server/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java

```diff
@@ -109,26 +111,24 @@ protected NodeGatewayStartedShards nodeOperation(NodeRequest request, Task task)
             ShardStateMetaData shardStateMetaData = ShardStateMetaData.FORMAT.loadLatestState(logger, namedXContentRegistry,
                 nodeEnv.availableShardPaths(request.shardId));
             if (shardStateMetaData != null) {
-                IndexMetaData metaData = clusterService.state().metaData().index(shardId.getIndex());
-                if (metaData == null) {
-                    // we may send this requests while processing the cluster state that recovered the index
-                    // sometimes the request comes in before the local node processed that cluster state
-                    // in such cases we can load it from disk
-                    metaData = IndexMetaData.FORMAT.loadLatestState(logger, namedXContentRegistry,
```

Comment:
Are we good with not having this fallback in a mixed version cluster?
If so (it seems we're fixing quite the edge case here), couldn't we get rid of the extra disk read by just waiting for the cluster state to have this index via a CS observer without having to add redundant (since we already sent that same inforamtion via the CS) information to the transport request?

Reply:
I wanted to avoid using `ClusterStateObserver` as it has the potential of blocking shard allocation for a while  (One slow node applying the cluster state very slowly), and also it requires making this method async. Instead I opted to throw an exception in that case, which will have the master retry the search for shards, and which is simpler in this case.

Reply:
Makes sense, thanks for explaining :)